### PR TITLE
WIP: Implement new interface

### PR DESCRIPTION
**Pull request by nmheim**
described in #21 

- [ ] use `softplus_safe`
- [ ] Dirac and CDirac
- [ ] Mixture
- [ ] FromDistribution (thin wrapper of our interface to general class of Distributions)
- [x] doc strings
- [x] implement shared variance `CGaussian` (what to call it?)
- [x] add readme
- [x] Gaussian and CGaussian
- [x] are we happy with the name `Gaussian`? Distributions uses `Normal` but calling our normal distributions like theirs results in amiguities with functions like `mean(p::Normal)`
- [x] what should `length(p::CPDF)` return if it should be a function at all?
- [x] why do all other `length`s dissapear form global namespace if I `export length`?

### CODE REVIEW

**Comment by vitskvara** [comment]
The new interface is nice, however instead of 
```julia
variance(p::CGaussian, z::AbstractArray) = p.σ2
```
we should have

```julia
variance(p::CGaussian, z::AbstractArray) = p.σ2(z)
```
just like with mean. Then we should merge it with the new types for output variance of VAE and set the mean and sigma2 in the pdf in the correct constructor.

**Comment by nmheim** [comment]
I think we could stick the diag/scalar/unit variances in the pdf types. started to implement it in the commit above (only for the `Gaussian` type). @vitskvara @smidl  what do you think?

(The 0-dim arrays are nonsense. Imagine that would be a scalar)

**Comment by vitskvara** [comment]
I am not sure that it is needed here, when we can simply define different loglikelihoods for different combinations of means and sigmas (matrix and matrix, matrix and scalar, matrix and vector). 

Why I implement these three variants is because of the conditional pdf, where you define the extraction of variance in different ways. There I am still missing the possibility of sigma being a function of x/z.

**Comment by vitskvara** [comment]
Why does the cgaussian have the field encoder? Since it is going to be used both for decoder and encoder, this does not make much sense.

**Comment by nmheim** [comment]
I didn't know what else to call it for now. It could be just `func` or something like that?

File: Project.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
 DrWatson = "634d3b9d-ee7a-5ddf-bec9-22491ea816e1"
 Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
```
Comment (nmheim) [inline Project.toml]: I think we don't depend on `Parameters.jl` except the tests or did I miss something? Thats why I only added them to the test dependencies

Comment (nmheim) [reply Project.toml]: sorry for being such a nitpick ^^

**Comment by vitskvara** [comment]
Julia complained that they are missing from the project dependency, so I
automatically added them. Feel free to remove them.

Dne po 16. 9. 2019 15:44 uživatel Niklas Heim <notifications@github.com>
napsal:

**Pull request merged**